### PR TITLE
Improve BoundedElasticScheduler to be less blocking

### LINKED ISSUE #2137: BoundedElasticScheduler should not block on submit/dispose
`PriorityBlockingQueue` is used, but it is accessed from non-blocking scopes (like `schedule` or `dispose`)

Context:
https://gitter.im/reactor/BlockHound?at=5ea3064e94ac2758e3065670

### PR DESCRIPTION
**Pull request by simonbasle**
- Add BlockHound exception for BoundedElasticScheduler ensureQueueCapacity
- Remove PriorityQueue in BoundedElasticScheduler, use array+iteration

Fixes #2137.

### CODE REVIEW

**Comment by simonbasle** [comment]
note that I wasn't really able to trigger the blocking behavior and BlockHound complaint around the `PriorityQueue`.
I also need to perform some perf measurement (even maybe commit a jmh bench) to see the impact of that change.

**Comment by simonbasle** [comment]
I need to refine the benchmark, but preliminary benchmark of running 1000 tasks on a 100 thread bounded scheduler shows the following results:

```
Benchmark                                               Mode  Cnt   Score   Error  Units
BoundedElasticBusyStructureBenchmark.withArray          avgt   50  19.148 ± 2.504  ms/op
BoundedElasticBusyStructureBenchmark.withPriorityQueue  avgt   50  29.360 ± 0.731  ms/op
```

**Comment by simonbasle** [comment]
I'm not entirely comfortable delivering this in next week's release without some user feedback.
@ssalamov would you be able to build a snapshot locally with this branch and validate that all is working well for you ?

**Comment by OlegDokuka** [comment]
we may need a stresstest for the updated impl.

**Comment by OlegDokuka** [comment]
Looks good to me in general. Curious if it is possible to see the efficiency of the algorithm in case of tasks which are heterogeneous in time

**Review by OlegDokuka — APPROVED** [review]

File: reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java [hunk 1]
```diff
@@ -402,30 +472,45 @@ else if (a < parent.maxThreads) {
...
+				return null; //implicit retry in the pick() loop
 			}
+			if (len == 1) {
+				return arr[0];
+			}
+
+			BoundedState choice = arr[0];
+			int leastBusy = Integer.MAX_VALUE;
+
+			for (int i = 0; i < arr.length; i++) {
```
Comment (OlegDokuka) [inline reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java:505]: can we do random choices here instead of iterating?

Comment (OlegDokuka) [reply reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java:505]: that is what we do in rsocket weighted loadbalancer -> https://github.com/rsocket/rsocket-java/blob/master/rsocket-core/src/main/java/io/rsocket/loadbalance/WeightedLoadbalanceStrategy.java#L91

**Comment by simonbasle** [comment]
@ssalamov it should, yes, as it will entirely remove the `PriorityBlockingQueue` from the implementation

**Comment by ssalamov** [comment]
@simonbasle, I tested and couldn't find the blocking operation with `BoundedElasticScheduler`. I would like to confirm this fix works ;) Thank you for handling it.

File: reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java (hunk 1 above)
Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java:505]: let me capture that as a follow-up issue and merge this one for now 👍

Comment (simonbasle) [reply reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java:505]: see #2927

**Pull request merged**